ACME webhook conformance tests crashing with cert-manager 1.11.0 and 1.10.2

<!--
Bugs should be filed for issues encountered whilst operating cert-manager.
You should first attempt to resolve your issues through the community support
channels, e.g. Slack, in order to rule out individual configuration errors.
Please provide as much detail as possible. 
-->

**Describe the bug**:
After upgrading the Go dependency `github.com/cert-manager/cert-manager` from `1.10.1` to `1.10.2` or `1.11.0` in an ACME webhook the conformance tests are crashing.

The following output is generated by `go test -v ./test`:
```
=== RUN   TestRunsSuite
    fixture.go:120: unknown solver type: hetzner
--- FAIL: TestRunsSuite (2.88s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x172258b]

goroutine 24 [running]:
testing.tRunner.func1.2({0x1910820, 0x2b86230})
        /usr/local/go/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
        /usr/local/go/src/testing/testing.go:1399 +0x39f
panic({0x1910820, 0x2b86230})
        /usr/local/go/src/runtime/panic.go:884 +0x212
github.com/cert-manager/cert-manager/test/acme/dns.(*fixture).setup(0xc000125380, 0xc000417860)
        /go/pkg/mod/github.com/cert-manager/cert-manager@v1.11.0/test/acme/dns/fixture.go:123 +0x42b
github.com/cert-manager/cert-manager/test/acme/dns.(*fixture).RunConformance(0xc000125380, 0x7?)
        /go/pkg/mod/github.com/cert-manager/cert-manager@v1.11.0/test/acme/dns/fixture.go:134 +0x38
github.com/vadimkim/cert-manager-webhook-hetzner/test.TestRunsSuite(0x0?)
        /workspace/test/main_test.go:32 +0x1bf
testing.tRunner(0xc000417860, 0x1c686d8)
        /usr/local/go/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:1493 +0x35f
FAIL    github.com/vadimkim/cert-manager-webhook-hetzner/test   2.903s
FAIL
```

**Expected behaviour**:
ACME webhook tests which ran successfully with cert-manager 1.10.1 should also run successfully with 1.10.2 and 1.11.0

**Steps to reproduce the bug**:
- Update an ACME webhook to use cert-manager 1.10.2 or 1.11.0
- Adjust an ACME webhook conformance test to the new format in 1.10.2 / 1.11.0 by replacing the solver instance by its name
- Execute an ACME webhook conformance test.

```go
package test

import (
	"github.com/vadimkim/cert-manager-webhook-hetzner/internal"
	"math/rand"
	"os"
	"testing"

	"github.com/cert-manager/cert-manager/test/acme/dns"
)

var (
	zone = os.Getenv("TEST_ZONE_NAME")
	fqdn string
)

func TestRunsSuite(t *testing.T) {
	fqdn = GetRandomString(20) + "." + zone

	fixture := dns.NewFixture(
		internal.SolverName,
		dns.SetResolvedZone(zone),
		dns.SetResolvedFQDN(fqdn),
		dns.SetAllowAmbientCredentials(false),
		dns.SetManifestPath("testdata/hetzner"),
	)

	fixture.RunConformance(t)
}
```

**Anything else we need to know?**:
This is most pobably caused by the commits https://github.com/cert-manager/cert-manager/commit/8ed0faf2287c703cb32b93170bd51fa6994bb89d in 1.11.0 and https://github.com/cert-manager/cert-manager/commit/14f72bafb97ff1a27aab57c8af53e0192b6ffb66 in 1.10.2.

The offending code is here: https://github.com/cert-manager/cert-manager/blob/v1.10.2/test/acme/dns/fixture.go#L104-L123
The changes those commits introduced prevent all but the RFC2136 solver to be tested.
Because the code does not abort in case an "invalid" solver name was given the `testSolver` is never set, but the execution continues.

**Environment details:**:
- Kubernetes version: 1.26.0
- Cloud-provider/provisioner: n/a
- cert-manager version: 1.10.2 / 1.11.0
- Install method: Go module

/kind bug

--- a/pkg/acme/webhook/apiserver/apiserver.go
+++ b/pkg/acme/webhook/apiserver/apiserver.go
@@ -112,7 +112,10 @@ func (c *Config) Complete() CompletedConfig {
 	return CompletedConfig{&completedCfg}
 }
 
-// New returns a new instance of AdmissionServer from the given config.
+// New returns a new instance of apiserver from the given config. Each of the
+// configured solvers will have an API GroupVersion registered with the new
+// apiserver and will have its Initialize function passed as post-start hook
+// with the server.
 func (c completedConfig) New() (*ChallengeServer, error) {
 	genericServer, err := c.GenericConfig.New("challenge-server", genericapiserver.NewEmptyDelegate()) // completion is done in Complete, no need for a second time
 	if err != nil {

--- a/pkg/acme/webhook/cmd/cmd.go
+++ b/pkg/acme/webhook/cmd/cmd.go
@@ -29,6 +29,11 @@ import (
 	logf "github.com/cert-manager/cert-manager/pkg/logs"
 )
 
+// RunWebhookServer creates and starts a new apiserver that acts as a external
+// webhook server for solving DNS challenges using the provided solver
+// implementations. This can be used as an entry point by external webhook
+// implementations, see
+// https://github.com/cert-manager/webhook-example/blob/899c408751425f8d0842b61c0e62fd8035d00316/main.go#L23-L31
 func RunWebhookServer(groupName string, hooks ...webhook.Solver) {
 	stopCh, exit := util.SetupExitHandler(util.GracefulShutdown)
 	defer exit() // This function might call os.Exit, so defer last

--- a/pkg/acme/webhook/cmd/server/start.go
+++ b/pkg/acme/webhook/cmd/server/start.go
@@ -97,6 +97,9 @@ func (o *WebhookServerOptions) Complete() error {
 	return nil
 }
 
+// Config creates a new webhook server config that includes generic upstream
+// apiserver options, rest client config and the Solvers configured for this
+// webhook server
 func (o WebhookServerOptions) Config() (*apiserver.Config, error) {
 	// TODO have a "real" external address
 	if err := o.RecommendedOptions.SecureServing.MaybeDefaultWithSelfSignedCerts("localhost", nil, []net.IP{net.ParseIP("127.0.0.1")}); err != nil {
@@ -118,6 +121,8 @@ func (o WebhookServerOptions) Config() (*apiserver.Config, error) {
 	return config, nil
 }
 
+// RunWebhookServer creates a new apiserver, registers an API Group for each of
+// the configured solvers and runs the new apiserver.
 func (o WebhookServerOptions) RunWebhookServer(stopCh <-chan struct{}) error {
 	config, err := o.Config()
 	if err != nil {

--- a/pkg/acme/webhook/webhook.go
+++ b/pkg/acme/webhook/webhook.go
@@ -24,7 +24,9 @@ import (
 	whapi "github.com/cert-manager/cert-manager/pkg/acme/webhook/apis/acme/v1alpha1"
 )
 
-// Solver has the functionality to solve ACME challenges.
+// Solver has the functionality to solve ACME challenges. This interface is
+// implemented internally by RFC2136 DNS provider and by external webhook solver
+// implementations see https://github.com/cert-manager/webhook-example
 type Solver interface {
 	// Name is the name of this ACME solver as part of the API group.
 	// This must match what you configure in the ACME Issuer's DNS01 config.
@@ -41,5 +43,6 @@ type Solver interface {
 	CleanUp(ch *whapi.ChallengeRequest) error
 
 	// Initialize is called as a post-start hook when the apiserver starts.
+	// https://github.com/kubernetes/apiserver/blob/release-1.26/pkg/server/hooks.go#L32-L42
 	Initialize(kubeClientConfig *restclient.Config, stopCh <-chan struct{}) error
 }

--- a/pkg/issuer/acme/dns/rfc2136/provider.go
+++ b/pkg/issuer/acme/dns/rfc2136/provider.go
@@ -37,6 +37,8 @@ const SolverName = "rfc2136"
 
 type Solver struct {
 	secretLister corelisters.SecretLister
+	// options to apply when the lister gets initialized
+	initOpts []Option
 
 	// If specified, namespace will cause the rfc2136 provider to limit the
 	// scope of the lister/watcher to a single namespace, to allow for
@@ -55,6 +57,21 @@ func WithNamespace(ns string) Option {
 func WithSecretsLister(secretLister corelisters.SecretLister) Option {
 	return func(s *Solver) {
 		s.secretLister = secretLister
+	}
+}
+
+// InitializeResetLister is a hack to make RFC2136 solver fit the Solver
+// interface. Unlike external solvers that are run as apiserver implementations,
+// this solver is created as part of challenge controller initialization. That
+// makes its Initialize method not fit the Solver interface very well as we want
+// a way to initialize the solver with the existing Secrets lister rather than a
+// new kube apiserver client. InitializeResetLister allows to reset secrets
+// lister when Initialize function is called so that a new lister can be
+// created. This is useful in tests where a kube clientset can get recreated for
+// an existing solver (which would not happen when this solver runs normally).
+func InitializeResetLister() Option {
+	return func(s *Solver) {
+		s.initOpts = []Option{func(s *Solver) { s.secretLister = nil }}
 	}
 }
 
@@ -99,12 +116,12 @@ func (s *Solver) CleanUp(ch *whapi.ChallengeRequest) error {
 }
 
 func (s *Solver) Initialize(kubeClientConfig *restclient.Config, stopCh <-chan struct{}) error {
+	for _, opt := range s.initOpts {
+		opt(s)
+	}
 	// Only start a secrets informerfactory if it is needed (if the solver
 	// is not already initialized with a secrets lister) This is legacy
-	// functionality. If you have a secrets watcher already available in the
-	// caller, you probably want to use that to avoid double caching the
-	// Secrets
-	// TODO: refactor and remove this functionality
+	// functionality and is currently only used in integration tests.
 	if s.secretLister == nil {
 		cl, err := kubernetes.NewForConfig(kubeClientConfig)
 		if err != nil {

--- a/test/acme/dns/fixture.go
+++ b/test/acme/dns/fixture.go
@@ -81,6 +81,32 @@ type fixture struct {
 	propagationLimit time.Duration
 }
 
+// RunConformance will execute all conformance tests using the supplied
+// configuration These conformance tests should be run by all external DNS
+// solver webhook implementations, see
+// https://github.com/cert-manager/webhook-example
+func (f *fixture) RunConformance(t *testing.T) {
+	defer f.setup(t)()
+	t.Run("Conformance", func(t *testing.T) {
+		f.RunBasic(t)
+		f.RunExtended(t)
+	})
+}
+
+func (f *fixture) RunBasic(t *testing.T) {
+	defer f.setup(t)()
+	t.Run("Basic", func(t *testing.T) {
+		t.Run("PresentRecord", f.TestBasicPresentRecord)
+	})
+}
+
+func (f *fixture) RunExtended(t *testing.T) {
+	defer f.setup(t)()
+	t.Run("Extended", func(t *testing.T) {
+		t.Run("DeletingOneRecordRetainsOthers", f.TestExtendedDeletingOneRecordRetainsOthers)
+	})
+}
+
 func (f *fixture) setup(t *testing.T) func() {
 	f.setupLock.Lock()
 	defer f.setupLock.Unlock()
@@ -126,28 +152,4 @@ func (f *fixture) setup(t *testing.T) func() {
 		close(stopCh)
 		stopFunc()
 	}
-}
-
-// RunConformance will execute all conformance tests using the supplied
-// configuration
-func (f *fixture) RunConformance(t *testing.T) {
-	defer f.setup(t)()
-	t.Run("Conformance", func(t *testing.T) {
-		f.RunBasic(t)
-		f.RunExtended(t)
-	})
-}
-
-func (f *fixture) RunBasic(t *testing.T) {
-	defer f.setup(t)()
-	t.Run("Basic", func(t *testing.T) {
-		t.Run("PresentRecord", f.TestBasicPresentRecord)
-	})
-}
-
-func (f *fixture) RunExtended(t *testing.T) {
-	defer f.setup(t)()
-	t.Run("Extended", func(t *testing.T) {
-		t.Run("DeletingOneRecordRetainsOthers", f.TestExtendedDeletingOneRecordRetainsOthers)
-	})
 }

--- a/test/acme/dns/fixture.go
+++ b/test/acme/dns/fixture.go
@@ -24,12 +24,10 @@ import (
 	"time"
 
 	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
-	"k8s.io/client-go/informers"
 	"k8s.io/client-go/kubernetes"
 	"sigs.k8s.io/controller-runtime/pkg/envtest"
 
 	"github.com/cert-manager/cert-manager/pkg/acme/webhook"
-	"github.com/cert-manager/cert-manager/pkg/issuer/acme/dns/rfc2136"
 	"github.com/cert-manager/cert-manager/test/internal/apiserver"
 )
 
@@ -44,9 +42,7 @@ func init() {
 type fixture struct {
 	// testSolver is the actual DNS solver that is under test.
 	// It is set when calling the NewFixture function.
-	testSolver     webhook.Solver
-	testSolverType string
-
+	testSolver              webhook.Solver
 	resolvedFQDN            string
 	resolvedZone            string
 	allowAmbientCredentials bool
@@ -126,27 +122,7 @@ func (f *fixture) setup(t *testing.T) func() {
 
 	stopCh := make(chan struct{})
 
-	var testSolver webhook.Solver
-	switch f.testSolverType {
-	case rfc2136.SolverName:
-		cl, err := kubernetes.NewForConfig(env.Config)
-		if err != nil {
-			t.Errorf("error initializing solver: %#+v", err)
-		}
-
-		// obtain a secret lister and start the informer factory to populate the
-		// secret cache
-		factory := informers.NewSharedInformerFactoryWithOptions(cl, time.Minute*5)
-		secretLister := factory.Core().V1().Secrets().Lister()
-		factory.Start(stopCh)
-		factory.WaitForCacheSync(stopCh)
-		testSolver = rfc2136.New(rfc2136.WithSecretsLister(secretLister))
-		f.testSolver = testSolver
-	default:
-		t.Errorf("unknown solver type: %s", f.testSolverType)
-	}
-
-	testSolver.Initialize(env.Config, stopCh)
+	f.testSolver.Initialize(env.Config, stopCh)
 
 	return func() {
 		close(stopCh)

--- a/test/acme/dns/options.go
+++ b/test/acme/dns/options.go
@@ -23,8 +23,9 @@ import (
 	"strings"
 	"time"
 
-	"github.com/cert-manager/cert-manager/pkg/acme/webhook"
 	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
+	"github.com/cert-manager/cert-manager/pkg/acme/webhook"
 )
 
 // Option applies a configuration option to the test fixture being built

--- a/test/acme/dns/options.go
+++ b/test/acme/dns/options.go
@@ -23,6 +23,7 @@ import (
 	"strings"
 	"time"
 
+	"github.com/cert-manager/cert-manager/pkg/acme/webhook"
 	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 )
 
@@ -30,10 +31,13 @@ import (
 type Option func(*fixture)
 
 // NewFixture constructs a new *fixture, applying the given Options before
-// returning.
-func NewFixture(solverType string, opts ...Option) *fixture {
+// returning. Solver is an implementation of
+// https://github.com/cert-manager/cert-manager/blob/v1.11.0/pkg/acme/webhook/webhook.go#L27-L45
+// and could be RFC2136 solver or any of external solvers that run these
+// conformance tests.
+func NewFixture(solver webhook.Solver, opts ...Option) *fixture {
 	f := &fixture{
-		testSolverType: solverType,
+		testSolver: solver,
 	}
 	for _, o := range opts {
 		o(f)

--- a/test/integration/rfc2136_dns01/provider_test.go
+++ b/test/integration/rfc2136_dns01/provider_test.go
@@ -59,7 +59,7 @@ func TestRunSuiteWithTSIG(t *testing.T) {
 		TSIGKeyName: rfc2136TestTsigKeyName,
 	}
 
-	fixture := dns.NewFixture(rfc2136.SolverName,
+	fixture := dns.NewFixture(rfc2136.New(rfc2136.InitializeResetLister()),
 		dns.SetResolvedZone(rfc2136TestZone),
 		dns.SetResolvedFQDN(rfc2136TestFqdn),
 		dns.SetAllowAmbientCredentials(false),
@@ -91,7 +91,7 @@ func TestRunSuiteNoTSIG(t *testing.T) {
 		Nameserver: server.ListenAddr(),
 	}
 
-	fixture := dns.NewFixture(rfc2136.SolverName,
+	fixture := dns.NewFixture(rfc2136.New(rfc2136.InitializeResetLister()),
 		dns.SetResolvedZone(rfc2136TestZone),
 		dns.SetResolvedFQDN(rfc2136TestFqdn),
 		dns.SetAllowAmbientCredentials(false),

